cpp/src/arrow/array/array_nested.cc

```diff
@@ -678,6 +680,10 @@ Result<std::shared_ptr<Array>> DenseUnionArray::Make(
     return Status::TypeError("UnionArray type_ids must be signed int8");
   }
 
+  if (type_ids.null_count() != 0) {
+    return Status::Invalid("Union type ids may not have nulls");
+  }
+
   if (value_offsets.null_count() != 0) {
     return Status::Invalid("Make does not allow NAs in value_offsets");
```

Comment:
Slightly OT, but shouldn't we normalize "NA" vs "null" in our error messages?

Reply:
Yes, I'll search for usages of "NA" and replace with null